Adding FVFunctorNeumannBC and FVFunctorConvectiveHeatFluxBC

coopertrucks opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

@snschune 
## Reason
Necessary to transfer functors between multiapps, particularly from finite element to finite volume

## Design
New classes added

## Impact
None


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

Commit:
Add FVFunctorNeumannBC

Commit:
Adding documentation and test for FVFunctorNeumannBC (#21632)

Commit:
Added FVFunctorConvectiveHeatFluxBC and revised FVFunctorNeumannBC test (#21632)

Commit:
Added FVFunctorConvectiveHeatFluxBC test and documentation for FVFunctorConvectiveHeatFluxBC and FVFunctorNeumannBC (#21632)

GiudGiud commented on modules/navier_stokes/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,19 @@
+For a fluid domain, the negative of the heat flux is applied. This allows for easier implementation of a double Robin
+boundary condition.
+
+See [CoupledConvectiveHeatFluxBC](CoupledConvectiveHeatFluxBC.md) for a similar boundary condition coupled to variables.
```
also add FVFunctorThermalResistance, which actually already does conduction + convection + radiation

GiudGiud commented on modules/navier_stokes/include/fvbcs/FVFunctorConvectiveHeatFluxBC.h:
```diff
@@ -0,0 +1,39 @@
+#include "FVFluxBC.h"
+
+/**
+ * Neumann boundary (== inflow) condition for finite volume scheme
+ * where the inflow is given as a functor
```
update this

GiudGiud commented on modules/navier_stokes/include/fvbcs/FVFunctorConvectiveHeatFluxBC.h:
```diff
@@ -0,0 +1,39 @@
+protected:
+  virtual ADReal computeQpResidual() override;
+
+  /// Walltemperature variable
```
they are not variables but functors

GiudGiud commented on modules/navier_stokes/include/fvbcs/FVFunctorConvectiveHeatFluxBC.h:
```diff
@@ -0,0 +1,39 @@
+  /// Convective heat transfer coefficient
+  const Moose::Functor<ADReal> & _htc;
+
+  const bool _is_solid;
+  // unsigned int _t
```
clean up + docstrings

GiudGiud commented on modules/navier_stokes/src/fvbcs/FVFunctorConvectiveHeatFluxBC.C:
```diff
@@ -0,0 +1,61 @@
+  params.addRequiredParam<MooseFunctorName>("heat_transfer_coefficient",
+                                                "Functor for heat transfer coefficient");
+  params.addRequiredParam<bool>("is_solid", "Whether this kernel acts on the solid temperature");
+  // params.addParam<MaterialPropertyName>(
+  //     "heat_transfer_coefficient_dT",
+  //     "0",
+  //     "Material property for derivative of heat transfer coefficient with respect to temperature");
```
```suggestion
```

GiudGiud commented on modules/navier_stokes/src/fvbcs/FVFunctorNeumannBC.C:
```diff
@@ -0,0 +1,38 @@
+{
+  InputParameters params = FVFluxBC::validParams();
+  params.addClassDescription("Neumann boundary condition for finite volume method.");
+  params.addParam<Real>("factor",
+                        1.,
```
make the factor a functor. It ll come in handy one day

GiudGiud reviewed (changes requested):
The heat convection BC should go in modules/heat_conduction probably. This is the heat_transfer module, to rename one day. 

Please note we already have a convection+radiation+conduction combined BC, it just uses a constant rather than a functor. You may want to also enable functors for that one.

coopertrucks commented:
> The heat convection BC should go in modules/heat_conduction probably. This is the heat_transfer module, to rename one day.
> 
> Please not we already have a convection+radiation+conduction combined BC, it just uses a constant rather than a functor. You may want to also enable functors for that one.

Moved to modules/heat_conduction and the combined BC is noted. I will attempt to enable functors for it in the future once the other boundary conditions are good.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1119052/) on e93959e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21668/site)

This comment will be updated on new commits.

snschune commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,20 @@
+
+# Description
+
+This boundary condition computes convective heat flux $q'' = H \cdot (T - T_{inf})$, where $H$ is convective heat transfer coefficient,
```
I think we refer to it as lower case `h` usually....being nit-picky

snschune commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,20 @@
+
+This boundary condition computes convective heat flux $q'' = H \cdot (T - T_{inf})$, where $H$ is convective heat transfer coefficient,
+$T$ is the temperature, and $T_{inf}$ is far field temperature. Both $H$ and $T_{inf}$ are coupled as functors.
+The domain can be specified as either a fluid or a solid using $is_solid$. For a solid domain, the equation above is applied.
```
The domain of the `variable` can be specified ...(i.e. is the variable the solid or the fluid temperature).

snschune commented on modules/heat_conduction/include/fvbcs/FVFunctorConvectiveHeatFluxBC.h:
```diff
@@ -0,0 +1,40 @@
+/**
+ * Robin boundary condition (temperatures) for finite volume scheme between
+ * a solid and fluid where the temperatures and heat transfer coefficient
+ * is given as a functor
```
are given as functor, right?

snschune commented on modules/heat_conduction/src/fvbcs/FVFunctorConvectiveHeatFluxBC.C:
```diff
@@ -0,0 +1,49 @@
+ADReal
+FVFunctorConvectiveHeatFluxBC::computeQpResidual()
+{
+  if (_is_solid)
```
Some chime-in from computer-science educated people. 
Does it make a difference in runtime to either predefine `Real mult = is_solid ? -1 : 1` versus doing the `if (_is_solid)` every time this is called?

lindsayad replied:
The cost of the if condition check will be negligible compared to the functor evaluations that go on below this. But it would be nice to predefine the sign during construction and then this function could be a one-liner

Commit:
Further revisions and grammatical fixes (#21632)

coopertrucks commented:
Changes have been submitted by commit

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1123808/) on 2468e39 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d672f8</th>
      <th colspan=4>#21668 2468e3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.86%</td>
      <td>83.86%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78354</td>
      <td>78355</td>
      <td>+1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15081</td>
      <td>15080</td>
      <td>-1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21668/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21668/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>d672f8</th>
      <th colspan=4>#21668 2468e3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.75%</td>
      <td>89.75%</td>
      <td>+0.00%</td>
      <td>90.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3196</td>
      <td>3214</td>
      <td>+18</td>
      <td>18</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>365</td>
      <td>367</td>
      <td>+2</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21668/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21668/coverage/heat_conduction)

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>d672f8</th>
      <th colspan=4>#21668 2468e3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.47%</td>
      <td>85.48%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>9698</td>
      <td>9711</td>
      <td>+13</td>
      <td>13</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1649</td>
      <td>1649</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21668/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21668/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21668/coverage/xfem)

</details>



This comment will be updated on new commits.

snschune commented:
@cticenhour I don't understand the SQA error here. The files are where they should be but the error message indicates that they should be located in `framework/doc`.

GiudGiud reviewed (changes requested):
register the items to the respective HeatConductionApp and NavierStokesApp not MooseApp

this is the SQA failure you are seeing

snschune commented:
I think something went terribly wrong in the last push.

GiudGiud commented on modules/navier_stokes/src/fvbcs/FVFunctorNeumannBC.C:
```diff
@@ -0,0 +1,41 @@
+FVFunctorNeumannBC::validParams()
+{
+  InputParameters params = FVFluxBC::validParams();
+  params.addClassDescription("Neumann boundary condition for finite volume method.");
```
since this is in NS (probably could go in framework/) we should mention it cant be used for velocities

GiudGiud commented on modules/heat_conduction/test/tests/fvbcs/FVFunctorConvectiveHeatFluxBC/fv_functor_convective_heat_flux.i:
```diff
@@ -0,0 +1,78 @@
+[Mesh]
+  [gen]
+    type = GeneratedMeshGenerator
+    dim = 2
+    nx = 5
+    ny = 5
+    xmax = 2
+  []
```
rework your test to have multiple blocks. The T_wall and T_inf are likely meant to be defined on different sides of the interface

coopertrucks replied:
Reworked in next commit, FVFluxBC doesn't work on internal interfaces so I separated it into a multiapp with solid domain in one input and fluid domain in the other, with a nearest node transfer

GiudGiud replied:
it works on internal to the mesh, external to the domain of definition of a variable

GiudGiud commented on modules/heat_conduction/src/fvbcs/FVFunctorConvectiveHeatFluxBC.C:
```diff
@@ -0,0 +1,49 @@
+  params.addClassDescription(
+      "Convective heat transfer boundary condition with temperature and heat "
+      "transfer coefficent given by functors.");
+  params.addRequiredParam<MooseFunctorName>("T_wall", "Functor for wall temperature");
```
so wall is solid?
and infinity is fluid?

infinity is usually bulk btw

coopertrucks replied:
Yes, renamed in the next commit to T_solid and T_bulk

Commit:
Revised FVFunctorConvectiveHeatFluxBC test to include solid and fluid domain via multiapp, minor revisions (#21632)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1130528/) on 5ab5f83 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21668/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21668/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 004f3bd8409919b662f8e5020b65aa6c4bd9f40b`

GiudGiud commented:
I ll finish this

GiudGiud commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,20 @@
+
+# Description
+
+This boundary condition computes convective heat flux $q'' = h \cdot (T - T_{bulk})$, where $H$ is convective heat transfer coefficient,
```
```suggestion
This boundary condition computes convective heat flux $q'' = h \cdot (T - T_{bulk})$, where $h$ is the convective heat transfer coefficient,
```

GiudGiud commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,20 @@
+# Description
+
+This boundary condition computes convective heat flux $q'' = h \cdot (T - T_{bulk})$, where $H$ is convective heat transfer coefficient,
+$T$ is the temperature, and $T_{bulk}$ is far field temperature. Both $H$ and $T_{bulk}$ are coupled as functors.
```
```suggestion
$T$ is the temperature, and $T_{bulk}$ is the far-field temperature. Both $h$ and $T_{bulk}$ are functors, which enables various spatial, variable and other dependences.
```

GiudGiud commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,20 @@
+For a fluid domain, the negative of the heat flux is applied. This allows for easier implementation of a double Robin
+boundary condition.
+
+See [CoupledConvectiveHeatFluxBC](CoupledConvectiveHeatFluxBC.md) for a similar boundary condition coupled to variables, and see [FunctorThermalResistanceBC](FunctorThermalResistanceBC.md) for a combined conduction, convection, and radiative boundary condition
```
```suggestion
Similar objects:
- [CoupledConvectiveHeatFluxBC](CoupledConvectiveHeatFluxBC.md) for a similar boundary condition coupled to variables, for finite elements
- [FunctorThermalResistanceBC](FunctorThermalResistanceBC.md) for a combined conduction, convection, and radiative boundary condition, with a constant outside ambient temperature
```

GiudGiud commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,20 @@
+boundary condition.
+
+See [CoupledConvectiveHeatFluxBC](CoupledConvectiveHeatFluxBC.md) for a similar boundary condition coupled to variables, and see [FunctorThermalResistanceBC](FunctorThermalResistanceBC.md) for a combined conduction, convection, and radiative boundary condition
+ with functor material properties.
```
```suggestion
```

GiudGiud commented on modules/heat_conduction/src/fvbcs/FVFunctorConvectiveHeatFluxBC.C:
```diff
@@ -0,0 +1,49 @@
+  InputParameters params = FVFluxBC::validParams();
+  params.addClassDescription(
+      "Convective heat transfer boundary condition with temperature and heat "
+      "transfer coefficent given by functors.");
```
```suggestion
      "transfer coefficient given by functors.");
```

GiudGiud committed:
Rework FV convective heat flux test to be in a single app

GiudGiud committed:
Keep FVFunctorNeumannBC as a framework test (#21632)

GiudGiud commented:
@lindsayad for the review should this pass the tests. I ended up changing a few things but not very many.

lindsayad commented on framework/include/fvbcs/FVFunctorNeumannBC.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "FVFluxBC.h"
+
+class Function;
+
```
```suggestion
```

lindsayad commented on modules/heat_conduction/doc/content/source/fvbcs/FVFunctorConvectiveHeatFluxBC.md:
```diff
@@ -0,0 +1,28 @@
+
+This boundary condition computes convective heat flux $q'' = h \cdot (T - T_{bulk})$, where $h$ is the convective heat transfer coefficient,
+$T$ is the temperature, and $T_{bulk}$ is the far-field temperature. Both $h$ and $T_{bulk}$ are functors, which enables various spatial, variable and other dependences.
+The domain of the variable can be specified as either a fluid or a solid using $is\_solid$. For a solid domain, the equation above is applied.
```
```suggestion
The domain of the variable can be specified as either a fluid or a solid using the `is_solid` parameter. For a solid domain, the equation above is applied.
```

lindsayad commented on modules/heat_conduction/src/fvbcs/FVFunctorConvectiveHeatFluxBC.C:
```diff
@@ -0,0 +1,49 @@
+ADReal
+FVFunctorConvectiveHeatFluxBC::computeQpResidual()
+{
+  if (_is_solid)
+    return -_htc(singleSidedFaceArg()) *
+           (_T_bulk(singleSidedFaceArg()) - _T_solid(singleSidedFaceArg()));
+  else
+    return _htc(singleSidedFaceArg()) *
+           (_T_bulk(singleSidedFaceArg()) - _T_solid(singleSidedFaceArg()));
```
```suggestion
  const auto ssf = singleSidedFaceArg();
  const auto flux = htc(ssf) * (_T_bulk(ssf) - _T_solid(ssf));
  if (_is_solid)
    return -flux;
  else
    return flux;
```

GiudGiud committed:
Address Alex review

- remove an unnecessary declaration
- clean up flux computation for convective HT BC

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

GiudGiud commented:
Had to fix a suggestion need re-approval pls

GiudGiud commented:
All test failures unrelated (CO2 fp, sockeye, EM coverage)

GiudGiud commented:
@coopertrucks thank you for the contribution!
